src/cluster_resources.rs

```diff
@@ -0,0 +1,393 @@
+        &mut self,
+        client: &Client,
+        resource: &T,
+    ) -> OperatorResult<T> {
+        ClusterResources::check_label(resource.labels(), APP_INSTANCE_LABEL, &self.app_instance)?;
+        ClusterResources::check_label(resource.labels(), APP_MANAGED_BY_LABEL, &self.manager)?;
+        ClusterResources::check_label(resource.labels(), APP_NAME_LABEL, &self.app_name)?;
+
+        let patched_resource = client
+            .apply_patch(&self.manager, resource, resource)
+            .await?;
+
+        self.resources.insert((&patched_resource).into());
+
+        Ok(patched_resource)
+    }
+
+    /// Checks that the given `labels` contain the given `label` with the given `expected_content`.
+    ///
+    /// # Arguments
+    ///
+    /// * `labels` - The labels to check
+    /// * `label` - The expected label
+    /// * `expected_content` - The expected content of the label
+    ///
+    /// # Errors
+    ///
+    /// If `labels` does not contain `label` then an `Error::MissingLabel` is returned.
+    ///
+    /// If `labels` contains the given `label` but not with the `expected_content` then an
+    /// `Error::UnexpectedLabelContent` is returned
+    fn check_label(
+        labels: &BTreeMap<String, String>,
+        label: &'static str,
+        expected_content: &str,
+    ) -> OperatorResult<()> {
+        if let Some(actual_content) = labels.get(label) {
+            if expected_content == actual_content {
+                Ok(())
+            } else {
+                Err(Error::UnexpectedLabelContent {
+                    label,
+                    expected_content: expected_content.into(),
+                    actual_content: actual_content.into(),
+                })
+            }
+        } else {
+            Err(Error::MissingLabel { label })
+        }
+    }
+
+    /// Finalizes the cluster creation.
+    ///
+    /// All orphaned resources, i.e. resources which are labelled as if they belong to this cluster
+    /// instance but were not added to the cluster resources, are deleted.
+    ///
+    /// The following resource labels are compared:
+    /// * `app.kubernetes.io/instance`
+    /// * `app.kubernetes.io/managed-by`
+    /// * `app.kubernetes.io/name`
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    pub async fn finalize(self, client: &Client) -> OperatorResult<()> {
+        self.delete_orphaned_resources::<Service>(client).await?;
+        self.delete_orphaned_resources::<StatefulSet>(client)
+            .await?;
+        self.delete_orphaned_resources::<DaemonSet>(client).await?;
+        self.delete_orphaned_resources::<ConfigMap>(client).await?;
+
+        Ok(())
+    }
+
+    /// Deletes all deployed resources which are labelled as if they belong to this cluster
+    /// instance but are not contained in the given list.
+    ///
+    /// If it is forbidden to list the resources of the given kind then it is assumed that the
```

Comment:
Would it maybe be worthwhile to change the signature of this fn to return a OperatorResult<bool> or something similar, which then shows if resources were present that nothing could be done about?
Not sure what scenarios this might apply to, but if an operator later needs this functionality it might be easy to add now, but have a larger blast radius later.

Reply:
If a resource cannot be deleted then an error is returned. The function terminates after the first error. So if a list of resources, resource names, or errors should be returned then the `Error` enum must be extended accordingly, e.g.:

```rust
pub enum Error {
    #[error("Errors occurred: {errors:?}")]
    Errors { errors: Vec<Error> },
}
```

The `delete_orphaned_resources` function must then collect all failed resources, resource names, or results, e.g.:

```rust
let mut results = Vec::new();
for resource in orphaned_resources {
    let result = client.ensure_deleted(resource).await;
    results.push(result);
}
let errors = results
    .into_iter()
    .filter_map(Result::err)
    .collect::<Vec<_>>();
if errors.is_empty() {
    Ok(())
} else {
    Err(Error::Errors { errors })
}
```

Is this what you want to achieve?

Reply:
I think I misread your comment slightly..

> If it is forbidden to list the resources of the given kind then it is assumed that the caller is not in charge of these resources, the deletion is skipped, and no error is returned.

My thinking was that it simply skips some of the resources that it cannot delete, but it would simply skip the entire step when it is not allowed to list, which makes sense.
I was just wondering if the information that deletion was skipped might be interesting to the calling code at some point in time?

Reply:
> I was just wondering if the information that deletion was skipped might be interesting to the calling code at some point in time?

The `delete_orphaned_resouces` function tries to delete resources of the kinds which implement the `ClusterResource` trait. Currently there are four kinds which implement this trait. If an operator would treat skipped kinds and another kind is added because it is used in a different operator then this could break the operator trying to handle skipped kinds. From the perspective of one operator, the kinds handled by the `ClusterResources`, are arbitrary and therefore the operator should not handle them.

Reply:
makes sense